core/trino-main/src/main/java/io/trino/server/testing/exchange/LocalFileSystemExchangeSink.java

```diff
@@ -0,0 +1,147 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.server.testing.exchange;
+
+import io.airlift.log.Logger;
+import io.airlift.slice.OutputStreamSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceOutput;
+import io.airlift.units.DataSize;
+import io.trino.spi.exchange.ExchangeSink;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.io.MoreFiles.deleteDirectoryContents;
+import static com.google.common.io.RecursiveDeleteOption.ALLOW_INSECURE;
+import static io.airlift.units.DataSize.Unit.KILOBYTE;
+import static java.lang.Math.toIntExact;
+import static java.nio.file.Files.createFile;
+import static java.util.Objects.requireNonNull;
+
+public class LocalFileSystemExchangeSink
+        implements ExchangeSink
+{
+    private static final Logger log = Logger.get(LocalFileSystemExchangeSink.class);
+
+    private static final int BUFFER_SIZE_IN_BYTES = toIntExact(DataSize.of(4, KILOBYTE).toBytes());
+
+    private final Path outputDirectory;
+    private final int outputPartitionCount;
+
+    @GuardedBy("this")
+    private final Map<Integer, SliceOutput> outputs = new HashMap<>();
+    @GuardedBy("this")
+    private boolean committed;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public LocalFileSystemExchangeSink(Path outputDirectory, int outputPartitionCount)
+    {
+        this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+        this.outputPartitionCount = outputPartitionCount;
+    }
+
+    @Override
+    public synchronized void add(int partitionId, Slice data)
+    {
+        checkArgument(partitionId < outputPartitionCount, "partition id is expected to be less than %s: %s", outputPartitionCount, partitionId);
+        checkState(!committed, "already committed");
+        if (closed) {
+            return;
+        }
+        SliceOutput output = outputs.computeIfAbsent(partitionId, this::createOutput);
+        output.writeInt(data.length());
+        output.writeBytes(data);
+    }
+
+    private SliceOutput createOutput(int partitionId)
+    {
+        Path outputPath = outputDirectory.resolve(partitionId + ".data");
+        try {
+            return new OutputStreamSliceOutput(new FileOutputStream(outputPath.toFile()), BUFFER_SIZE_IN_BYTES);
+        }
+        catch (FileNotFoundException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public synchronized long getSystemMemoryUsage()
+    {
+        return BUFFER_SIZE_IN_BYTES * (long) outputs.size();
```

Comment:
You should probably iterate over buffers and use `SliceOutput.getRetainedSize` here.